[WFLY-11529] Expose WildFly metrics in Prometheus HTTP endpoint

jmesnil opened the pull request:
* add WildFly metrics (from subsystems and deployments) to the /metrics
  HTTP endpoint (only in the Prometheus text format)
* Use io.prometheus simpleclient 0.6.0 to collect the metrics and write
  them in TextFormat on the /metrics HTTP endpoint
* Name and labels of metrics are built based on the resource path
  address and the attribute name (plus an optional prefix that is set
  to `wildfly` in the standalone profiles).
* Update Admin guide

JIRA: https://issues.jboss.org/browse/WFLY-11529
Dev analysis: wildfly/wildfly-proposals#170

jmesnil committed:
[WFLY-11529] Expose WildFly metrics in Prometheus HTTP endpoint

* add WildFly metrics (from subsystems and deployments) to the /metrics
  HTTP endpoint (only in the Prometheus text format)
* Use io.prometheus simpleclient 0.6.0 to collect the metrics and write
  them in TextFormat on the /metrics HTTP endpoint
* Name and labels of metrics are built based on the resource path
  address and the attribute name (plus an optional prefix that is set
  to `wildfly` in the standalone profiles).
* Update Admin guide

JIRA: https://issues.jboss.org/browse/WFLY-11529

kabir commented:
@jmesnil @bstansberry to me this looks like a feature - please let's discuss on mail if you disagree with me

jmesnil commented:
@kabir you are right, this is the followup from https://issues.jboss.org/browse/WFLY-10712 (where WildFly metrics were disabled) that is tracked by 
https://issues.jboss.org/browse/EAP7-102.

I did not create a new EAP7 feature for it (as this PR effectively finishes the feature) but I should have added all the metadata to the PR description. I'm doing it right now

rsvoboda commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Metrics.adoc:
```diff
@@ -70,27 +71,27 @@ If the authentication fails, the  server will reply with a `401 NOT AUTHORIZED`
 The HTTP endpoint exposes the following metrics:
 
 * Base metrics - Required metrics specified in the MicroProfile 1.1 specification are exposed in the `base`  scope.
-* Vendor metrics - Metrics from WildFly subsystems are exposed in the `vendor` scope
```
Vendor metrics are there too

```
curl http://localhost:9990/metrics/vendor 2>/dev/null | grep -v TYPE | grep -v HELP | sort
vendor:buffer_pool_used_memory_direct_bytes 1519826.0
vendor:buffer_pool_used_memory_mapped_bytes 0.0
vendor:memory_pool_code_heap_non_nmethods_usage_bytes 1282944.0
vendor:memory_pool_code_heap_non_nmethods_usage_max_bytes 1345280.0
vendor:memory_pool_code_heap_non_profiled_nmethods_usage_bytes 2423552.0
vendor:memory_pool_code_heap_non_profiled_nmethods_usage_max_bytes 2423552.0
vendor:memory_pool_code_heap_profiled_nmethods_usage_bytes 9769984.0
vendor:memory_pool_code_heap_profiled_nmethods_usage_max_bytes 9769984.0
vendor:memory_pool_compressed_class_space_usage_bytes 8234296.0
vendor:memory_pool_compressed_class_space_usage_max_bytes 8234880.0
vendor:memory_pool_g1_eden_space_usage_bytes 3145728.0
vendor:memory_pool_g1_eden_space_usage_max_bytes 3.0408704E7
vendor:memory_pool_g1_old_gen_usage_bytes 4.7792496E7
vendor:memory_pool_g1_old_gen_usage_max_bytes 5.1015888E7
vendor:memory_pool_g1_survivor_space_usage_bytes 1048576.0
vendor:memory_pool_g1_survivor_space_usage_max_bytes 4194304.0
vendor:memory_pool_metaspace_usage_bytes 6.7490488E7
vendor:memory_pool_metaspace_usage_max_bytes 6.748984E7
```

jmesnil replied:
fixed

rsvoboda commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Metrics.adoc:
```diff
@@ -70,27 +71,27 @@ If the authentication fails, the  server will reply with a `401 NOT AUTHORIZED`
-
-WildFly metrics can appears in two scopes:
+The WildFly metrics (that measures activity in the subsystem or application deployments) are exposed only on the `/metrics`
+endpoint with the Prometheus format.
```
I can specify http://localhost:9990/metrics/base / http://localhost:9990/metrics/vendor / http://localhost:9990/metrics/application

Can I get just WildFly metrics ? Something like http://localhost:9990/metrics/wildfly ...

jmesnil replied:
no, you can't. WildFly metrics are only exposed in `/metrics`.
I see no advantage to provide an additional non-standard HTTP endpoint when Prometheus will just hit `/metrics` and get all the metrics (base/vendor/application + WildFly) in a single request.

rsvoboda commented on microprofile/metrics-smallrye/src/main/resources/org/wildfly/extension/microprofile/metrics/LocalDescriptions.properties:
```diff
@@ -1,5 +1,6 @@
+microprofile-metrics-smallrye.prefix=Prefix prepended to the name of the metrics exposed by the HTTP endpoints.
 microprofile-metrics-smallrye.remove=Remove the subsystem
 microprofile-metrics-smallrye.security-enabled=True if authentication is required to access the HTTP endpoint on the HTTP management interface.
 microprofile-metrics-smallrye.exposed-subsystems=The names of the subsystems that exposes their metrics in the vendor scope (or '*' to expose any subystem metrics).
```
This should be adjusted not to mention vendor scope

rsvoboda commented on microprofile/metrics-smallrye/src/main/resources/schema/wildfly-microprofile-metrics-smallrye_2_0.xsd:
```diff
@@ -0,0 +1,57 @@
+            <xs:attribute name="exposed-subsystems" type="xs:string">
+                <xs:annotation>
+                    <xs:documentation>
+                        The names of the subsystems (separated by spaces) that exposes their metrics in the vendor scope (or '*' to expose any subystem metrics).
```
This should be adjusted not to mention vendor scope

jmesnil replied:
fixed

rsvoboda commented on microprofile/metrics-smallrye/src/main/resources/subsystem-templates/microprofile-metrics-smallrye.xml:
```diff
@@ -3,8 +3,8 @@
-               exposed-subsystems="*">
-    </subsystem>
+               exposed-subsystems="*"
+               prefix="${wildfly.metrics.prefix:wildfly" />
```
Should EAP have its own prefix or not ?

jmesnil replied:
it's likely but that's not a concern for WildFly.

rsvoboda commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Metrics.adoc:
```diff
@@ -70,27 +71,27 @@ If the authentication fails, the  server will reply with a `401 NOT AUTHORIZED`
 
-* `vendor` for metrics defined in subsystem resources (e.g. the `bytes-received` attribute of the `/subsystem=undertow/server=default-server/http-listener=default` resource)
-* `applications` for metrics defined in deployment resources (e.g. the `request-count` attribute on the `/deployment=<deployment name>/subsystem=undertow/servlet=<servlet name>` resource).
+=== WildFly Metrics Description
```
WildFly Metrics is visible when accessing http://localhost:9990/metrics
WildFly Metrics is NOT visible when accessing http://localhost:9990/metrics/
The difference is trailing `/`

jmesnil replied:
yes, the endpoint is `/metrics` (but I'll patch it so that it also works with `/metrics/`)

jmesnil replied:
fixed

rsvoboda commented on feature-pack/pom.xml:
```diff
@@ -2892,6 +2892,15 @@
             </exclusions>
         </dependency>
 
+        <dependency>
```
60K	simpleclient/0.6.0/simpleclient-0.6.0.jar
8.0K	simpleclient_common/0.6.0/simpleclient_common-0.6.0.jar

Size is ok.
Which team own productization of those bits?
What's the sustaining/support plan for simpleclient ?

rsvoboda reviewed (changes requested):
Initial check, will look more into it during the week.

rsvoboda commented:
batch, ejb3, io, jca, request_controller, transactions are available when I hit `metrics` endpoint.
Undetrow is not there by default, I have to add `-Dwildfly.statistics-enabled=true` to see it in the output.

What's the difference of Undertow subsystem comparing to the others?

Invoking `curl http://localhost:9990/metrics 2>/dev/null | grep -v TYPE | grep -v HELP | sort`

jmesnil commented:
@rsvoboda some subsystems always provides their metrics (if they do not affect performance) such as `batch-jberet, `io`.

Others requires to set some `statistics-enabled` attribute to `true` to expose their metrics (`undertow`, `messaging-activemq`, etc.).
By defaultn these subsystems do not expose these metrics but we added the `wildfly.statistics-enabled` expression so that an user can easily expose them.

jmesnil committed:
[WFLY-11578] Distinguish WildFly metrics types

* In microprofile-metrics-smallrye extension, when attributes that
  corresponds to metrics are collected, check whether the attribute
  access has the COUNTER_METRIC flag to register it in Prometheus
  collector as a counter. Otherwise, the metric is registered as a
  gauge.
* Append _total to the Prometheus name for counters.

JIRA: https://issues.jboss.org/browse/WFLY-11578

jmesnil committed:
[WFLY-11578] Distinguish WildFly metrics

* Flag EJB3 metrics that represents counters with
  COUNTER_METRIC.

JIRA: https://issues.jboss.org/browse/WFLY-11578

jmesnil committed:
[WFLY-11578] Distinguish WildFly metrics

* Flag Transactions metrics that represents counters with
  COUNTER_METRIC.

JIRA: https://issues.jboss.org/browse/WFLY-11578

jmesnil committed:
[WFLY-11578] Distinguish WildFly metrics

* Flag messaging-activemq metrics that represents counters with
  COUNTER_METRIC.

JIRA: https://issues.jboss.org/browse/WFLY-11578

jmesnil committed:
[WFLY-11578] Distinguish WildFly metrics

* Flag Undertow metrics that represents counters with COUNTER_METRIC.

JIRA: https://issues.jboss.org/browse/WFLY-11578

rsvoboda commented:
https://github.com/jmesnil/wildfly/commit/496720849c4879e070acf639d2f22e7dd8957828 is there because of https://github.com/wildfly/wildfly-core/pull/3634 ?

jmesnil commented:
I have created a new EAP7 feature to track the exposure of WildFly metrics to avoid any confusion: https://issues.jboss.org/browse/EAP7-1169[

The dev analysis is at https://github.com/wildfly/wildfly-proposals/pull/170.

I've put this PR on hold as I want to add a tiny improvement (to distinguish counter from gauges) that requires an upgrade from WildFly Core (so that I can use new API from 
https://github.com/wildfly/wildfly-core/pull/3634).
This will be handled by an additional commit but the PR can be reviewed as it stands.

jmesnil commented:
I've updated the PR to include https://issues.jboss.org/browse/WFLY-11578 that is mentioned in the dev analysis so we properly distinguish counters from gauges.

rsvoboda reviewed (commented):
With latest changes you touched subsystems like EJB, Transactions, Undertow.

I didn't find changes for Webservices subsystem - https://github.com/wildfly/wildfly/blob/master/webservices/server-integration/src/main/java/org/jboss/as/webservices/dmr/WSEndpointMetrics.java#L57

rsvoboda commented:
Jeff Mesnil: great, thanks. re https://github.com/wildfly/wildfly/blob/master/webservices/server-integration/src/main/java/org/jboss/as/webservices/dmr/WSEndpointMetrics.java#L57, I have not looked at all metrics
Jeff Mesnil: I wanted to have the main (transactions, ejb, undertow) done for EAP7-1169, I'll open other PRs for these subsystems

Rostislav Svoboda:
```
git grep "AttributeDefinition" | grep -i COUNT | wc -l
      63
```
